File: packages/websearch/src/engines/mojeek.ts
import type {
  NamedWebSearchEngine,
  WebSearchEngineInput,
  WebSearchEngineResult,
  WebSearchResultItem,
} from "../types.js";
import { stripTags } from "./html.js";
import { httpGet } from "./http.js";
import { SearchError } from "./searchError.js";

const DEFAULT_BASE = "https://www.mojeek.com";
const ENGINE_NAME = "mojeek";

/**
 * Mojeek search — keyless HTML SERP parse, independent full-web crawl (not a
 * Google/Bing reseller), so it gives mainstream coverage the niche keyless
 * JSON APIs (Marginalia/Wikipedia) lack.
 *
 *   GET https://www.mojeek.com/search?q={query}
 *   → HTML; result list under <ul class="results-standard">, each result a
 *     block delimited by <!--rs--> ... <!--re--> containing
 *       <a class="title" href="URL">TITLE</a> ... <p class="s">SNIPPET</p>
 *
 * ToS note: Mojeek's robots.txt disallows /search and they sell an official
 * API — this is a scrape, a ToS gray area. It is therefore **opt-out**: the
 * fallback chain includes it by default for coverage, but a session can drop
 * it with `disableMojeek: true` (and the design doc flags it). We send our
 * honest agent User-Agent (no browser spoofing).
 *
 * @param opts.baseUrl override the host for tests (fixture server).
 */
export function createMojeekEngine(
  opts: { baseUrl?: string } = {},
): NamedWebSearchEngine {
  const base = opts.baseUrl ?? DEFAULT_BASE;
  return {
    name: ENGINE_NAME,
    engineClass: "general",
    async search(
      input: WebSearchEngineInput,
    ): Promise<WebSearchEngineResult> {
      const url = new URL(base);
      url.pathname = joinPath(url.pathname, "search");
      url.searchParams.set("q", input.query);

      const res = await httpGet(url, input, {
        accept: "text/html,application/xhtml+xml",
        engine: ENGINE_NAME,
      });

      // Distinguish a genuine "no hits" SERP from an anti-bot challenge.
      // A real Mojeek SERP — even with zero results — carries the result
      // scaffold (serp-results / results-count + "No pages found"); a
      // challenge/interstitial page has none of it. Only the latter should
      // fail the engine so the fallback chain moves on; a real empty SERP is
      // a legitimate `empty` outcome the chain may keep or surface.
      const results = parseMojeek(res.text).slice(0, input.count);
      if (results.length === 0 && looksChallenged(res.text)) {
        throw new SearchError(
          "SERVER_NOT_AVAILABLE",
          "mojeek returned no parseable results (likely an anti-bot challenge or interstitial from this IP)",
          { engine: ENGINE_NAME },
        );
      }
      return { results, backendHost: res.host, elapsedMs: res.elapsedMs };
    },
  };
}

/**
 * Parse Mojeek's result blocks. Exported for unit testing against a saved
 * fixture (no live network).
 */
export function parseMojeek(html: string): WebSearchResultItem[] {
  const out: WebSearchResultItem[] = [];
  const blockRe = /<!--rs-->([\s\S]*?)<!--re-->/g;
  let m: RegExpExecArray | null;
  while ((m = blockRe.exec(html)) !== null) {
    const block = m[1] ?? "";
    const titleMatch =
      /<a[^>]*class="title"[^>]*href="([^"]+)"[^>]*>([\s\S]*?)<\/a>/.exec(
        block,
      );
    if (!titleMatch) continue;
    const url = decodeHref(titleMatch[1] ?? "");
    const title = stripTags(titleMatch[2] ?? "");
    if (url.length === 0 || title.length === 0) continue;
    const snippetMatch = /<p class="s">([\s\S]*?)<\/p>/.exec(block);
    const snippet = snippetMatch ? stripTags(snippetMatch[1] ?? "") : "";
    out.push({ title, url, snippet });
  }
  return out;
}

function looksChallenged(html: string): boolean {
  // A genuine Mojeek SERP (even with zero hits) carries the result scaffold:
  // the "serp-results" container, a "results-count" bar, or an explicit
  // "No pages found" message. A bot-challenge / interstitial page has none of
  // these — that's what we treat as the engine being unavailable.
  const hasScaffold =
    html.includes("results-standard") ||
    html.includes("serp-results") ||
    html.includes("results-count") ||
    /no pages found/i.test(html);
  return !hasScaffold;
}

function decodeHref(href: string): string {
  return href.replace(/&amp;/g, "&");
}

function joinPath(basePath: string, segment: string): string {
  const trimmed = basePath.replace(/\/+$/, "");
  return `${trimmed}/${segment}`;
}

File: crates/websearch/src/engines/mojeek.rs
//! Mojeek search — keyless HTML SERP parse. See TS `engines/mojeek.ts`.
//! ToS note: Mojeek's robots.txt disallows /search and they sell an official
//! API; this is a scrape (gray area), so it is opt-out via the resolver.

use async_trait::async_trait;
use url::Url;

use super::html::strip_tags;
use super::http::http_get;
use crate::engine::{
    shared_client, SearchError, SearchErrorCode, WebSearchEngine, WebSearchEngineInput,
    WebSearchEngineResult,
};
use crate::types::{WebSearchResultItem, WebSearchTimeRange};

const DEFAULT_BASE: &str = "https://www.mojeek.com";
const ENGINE_NAME: &str = "mojeek";

pub struct MojeekEngine {
    client: reqwest::Client,
    base_url: String,
}

impl MojeekEngine {
    pub fn new() -> Self {
        Self {
            client: shared_client(),
            base_url: DEFAULT_BASE.to_string(),
        }
    }
    pub fn with_base_url(mut self, base: impl Into<String>) -> Self {
        self.base_url = base.into();
        self
    }
}

impl Default for MojeekEngine {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl WebSearchEngine for MojeekEngine {
    fn name(&self) -> &str {
        ENGINE_NAME
    }

    async fn search(
        &self,
        input: WebSearchEngineInput,
    ) -> Result<WebSearchEngineResult, SearchError> {
        let mut url = Url::parse(&self.base_url).map_err(|_| {
            SearchError::new(
                SearchErrorCode::IoError,
                format!("invalid mojeek base url: {}", self.base_url),
            )
        })?;
        {
            let base_path = url.path().trim_end_matches('/').to_string();
            url.set_path(&format!("{}/search", base_path));
            url.query_pairs_mut().append_pair("q", &input.query);
        }

        let res = http_get(
            &self.client,
            &url,
            &input,
            "text/html,application/xhtml+xml",
            ENGINE_NAME,
            &[],
        )
        .await?;

        let mut results = parse_mojeek(&res.text);
        results.truncate(input.count);

        // A real Mojeek SERP (even with zero hits) carries the result
        // scaffold; an anti-bot interstitial does not. Only the latter should
        // fail the engine so the fallback chain moves on.
        if results.is_empty() && looks_challenged(&res.text) {
            return Err(SearchError::new(
                SearchErrorCode::ServerNotAvailable,
                "mojeek returned no parseable results (likely an anti-bot challenge or interstitial from this IP)",
            ));
        }

        Ok(WebSearchEngineResult {
            results,
            backend_host: res.host,
            elapsed_ms: res.elapsed_ms,
            engine: Some(ENGINE_NAME.to_string()),
            engine_class: None,
            engines: None,
            // Mojeek's SERP scrape has no recency filter.
            time_range_applied: if input.time_range == WebSearchTimeRange::All {
                None
            } else {
                Some(false)
            },
        })
    }
}

/// Parse Mojeek's result blocks (delimited by `<!--rs-->` ... `<!--re-->`,
/// each with `<a class="title" href=...>` and `<p class="s">`). Public to the
/// crate for unit testing against a saved fixture.
pub(crate) fn parse_mojeek(html: &str) -> Vec<WebSearchResultItem> {
    let mut out = Vec::new();
    let mut rest = html;
    while let Some(start) = rest.find("<!--rs-->") {
        let after = &rest[start + "<!--rs-->".len()..];
        let end = match after.find("<!--re-->") {
            Some(e) => e,
            None => break,
        };
        let block = &after[..end];
        rest = &after[end + "<!--re-->".len()..];

        let (url, title) = match extract_title_anchor(block) {
            Some(t) => t,
            None => continue,
        };
        if url.is_empty() || title.is_empty() {
            continue;
        }
        let snippet = extract_snippet(block).unwrap_or_default();
        out.push(WebSearchResultItem {
            title,
            url,
            snippet,
            age: None,
            score: None,
            source: None,
        });
    }
    out
}

/// Find `<a ... class="title" ... href="URL" ...>TITLE</a>` and return
/// (decoded url, stripped title).
fn extract_title_anchor(block: &str) -> Option<(String, String)> {
    // Locate an <a ...> tag whose attributes include class="title".
    let mut search_from = 0;
    while let Some(rel) = block[search_from..].find("<a ") {
        let tag_start = search_from + rel;
        let tag_end = block[tag_start..].find('>')? + tag_start;
        let tag = &block[tag_start..=tag_end];
        if tag.contains("class=\"title\"") {
            let href = attr_value(tag, "href")?;
            // Title text is between this tag's '>' and the next "</a>".
            let after = &block[tag_end + 1..];
            let close = after.find("</a>")?;
            let title = strip_tags(&after[..close]);
            return Some((decode_href(&href), title));
        }
        search_from = tag_end + 1;
    }
    None
}

fn extract_snippet(block: &str) -> Option<String> {
    let marker = "<p class=\"s\">";
    let start = block.find(marker)? + marker.len();
    let after = &block[start..];
    let end = after.find("</p>")?;
    Some(strip_tags(&after[..end]))
}

/// Extract the value of `attr="..."` from a tag string.
fn attr_value(tag: &str, attr: &str) -> Option<String> {
    let needle = format!("{}=\"", attr);
    let start = tag.find(&needle)? + needle.len();
    let after = &tag[start..];
    let end = after.find('"')?;
    Some(after[..end].to_string())
}

fn decode_href(href: &str) -> String {
    href.replace("&amp;", "&")
}

fn looks_challenged(html: &str) -> bool {
    let has_scaffold = html.contains("results-standard")
        || html.contains("serp-results")
        || html.contains("results-count")
        || html.to_lowercase().contains("no pages found");
    !has_scaffold
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_blocks_with_title_url_snippet() {
        let html = r#"<ul class="results-standard">
<!--rs--><li class="r1"><a title="x" href="https://ex.com/a" class="ob"></a><h2><a class="title" href="https://ex.com/a">A Title</a></h2><p class="s">snippet <strong>a</strong></p></li><!--re-->
<!--rs--><li class="r2"><a class="title" href="https://ex.com/b">B Title</a><p class="s">snippet b</p></li><!--re-->
</ul>"#;
        let r = parse_mojeek(html);
        assert_eq!(r.len(), 2);
        assert_eq!(r[0].url, "https://ex.com/a");
        assert_eq!(r[0].title, "A Title");
        assert_eq!(r[0].snippet, "snippet a");
        assert_eq!(r[1].url, "https://ex.com/b");
    }

    #[test]
    fn empty_html_yields_no_results() {
        assert!(parse_mojeek("<html><body>nope</body></html>").is_empty());
    }

    #[test]
    fn decodes_ampersand_in_href() {
        let html = r#"<!--rs--><a class="title" href="https://ex.com/?a=1&amp;b=2">T</a><p class="s">s</p><!--re-->"#;
        let r = parse_mojeek(html);
        assert_eq!(r[0].url, "https://ex.com/?a=1&b=2");
    }
}
